### ISSUE TITLE
URL Redirection Infinite Loop

### ISSUE THREAD

**Issue by AndrewGlazier** [single-spa/single-spa-angular#113]
Issue seems to relate to this comment;
https://github.com/CanopyTax/single-spa-angular/issues/94#issuecomment-520550455
Creating a new ticket as I cannot reproduce using the browser back button.

This bug is reproducible in the vanilla example for coexisting angular micro-frontends found here:
https://github.com/joeldenning/coexisting-angular-microfrontends

Reproduction steps;
Using the navigation app, trigger a route change, and then quickly trigger another nav change.

Issue is reproducible on;
IE11 (Very common)
Firefox (Mediocre-Hard)
Chrome (Rare)

Notes;
In our application we have found that having more simultaneous angular apps running in parallel triggers the issue more commonly, however this bug is reproducible with only one app actively loaded by the single-spa (however with difficultly even in IE11). (Issue does not occur at all with the app running in angular standalone without the spa)

The issue appears to be also related to Router.forRoot in some way, upon removing this from an app, the issue becomes more difficult to reproduce even with IE11, a similar effect to whether the app wasn't running.

This seems to be related to the speed of the navigation. Navigating twice very quickly causes this effect.

If struggling to reproduce, we found that having a button which changes to swap between two pages and spam clicking it causes the bug to reproduce on IE11 every time.

The navigations do not need to be between separate apps, routing between two pages on the same app causes the issue.

**Comment by joeldenning** [comment]
I do not have time to work on this issue - outside contributions would be greatly appreciated.

**Comment by rknash18** [comment]
@AndrewGlazier I have fixed this issue by adding the onSameUrlNavigation: 'reload' option to router module in angular.

**Comment by joeldenning** [comment]
Oh wow thank you for sharing that @rknash18. It would be great if that workaround works for everyone else's issues they've been having, too. I've commented in #94 asking others to test if this works for them.

If it does, let's look into incorporating this into single-spa-angular so that all benefit from it.

**Comment by ashuaggarwal94** [comment]
Hi @joeldenning , @rknash18  ,

The solution provided in the comment does not work for me I m stilling using the changeDectection ref as an alternative.

**Comment by rknash18** [comment]
@ashuaggarwal94 I am not sure about the issue you are experiencing , I have fixed issue with router trying go back on forth when I clicked on a menu option.

**Comment by gubertcalixto** [comment]
I identified that this loop starts (in reroute.js file)
![image](https://user-images.githubusercontent.com/28396856/70262495-35d43900-1773-11ea-9954-95d6c4a6dd09.png)
In my case performAppChanges reached 400 times until the page was available again

**Comment by gubertcalixto** [comment]
My current guess is this [line](https://github.com/CanopyTax/single-spa/blob/master/src/navigation/navigation-events.js#L109). I think the second line
`urlReroute(createPopStateEvent(state))`
is runned after the second router change, causing unexpected behavior since browser is already changing again (to new url) and single-spa is processing the old url.
PS: **JUST A GUESS**, trying to debug it a little more.

**Comment by joeldenning** [comment]
Thanks for helping to debug this - why would single-spa's code have the old url? Angular router should call `replaceState` with the new url, not the old one.

**Comment by gubertcalixto** [comment]
What I meant was that single-spa reroute is happening during an router change, which causes the bug, but not sure about it at all, just like single spa is processing during another route change.

**Comment by gubertcalixto** [comment]
Test it out for about 2 hours. I'm creating an repo with the bug!
Notice that loop starts after "finishUpAndReturn" method is called twice, @joeldenning, any idea?

**Comment by gubertcalixto** [comment]
Created an [repo](https://github.com/gubertcalixto/single-spa-crazy-route) that reproduces the bug @joeldenning

**Comment by joeldenning** [comment]
I am not going to look at this bug anytime soon. I do not use Angular for any projects ever and this bug is a hard one that I tried in the past. The only path forward for this to be fixed is community contribution.

**Comment by gubertcalixto** [comment]
Notice something that might be the bug!!!
## Base Code
```typescript
this.router.navigateByUrl('threee').catch(res => {
    this.router.navigateByUrl('one');
});
```
## The Result
![image](https://user-images.githubusercontent.com/28396856/70857654-7baaa300-1ed1-11ea-8fec-01861b609779.png)
This image shows the arguments send to replaceState and to pushState.

## What I Found Out
1. As you can see, urls has **/test** prefix
- This is because I added `provide: APP_BASE_HREF, useValue: '/test'`, making all urls for that app to have /test prefix (/test will be in activtyFunction of SingleSPA)
2. What I'm interested in is that **first url** is /test/**two**, which means that Angular doesn't have changed route to **threee**.
- Actually, /test/two is the **current route in Angular App** (before redirecting to any route at all)
3. After that **Angular itself keep activating** replaceState and pushState, but only when singleSpa overrides this methods from window.history

**PS:** Finally found that if I do not set prefix (change from "/test" to "/"), the bug doesn't happen, even though Angular keeps sending "/two" other than "/threee"

**Comment by gubertcalixto** [comment]
Notice that the problem is in the following line
If I delete the following line, bug is fixed.
https://github.com/CanopyTax/single-spa/blob/3c17910b46f4b529a29eb850da70b97896ecb662/src/navigation/reroute.js#L118

@TheMcMurder (git blamed this code). I just need to know why is this code there. Because I not sure what might happen if I just delete it. And what you suggest, should I try to get why is pendingPromises making angular start a loop or just do something to fix this code (still don't know what)?

**Comment by TheMcMurder** [comment]
Full disclosure: I don't use angular for any projects.

I don't recall specifically why that line of code exists, but this is a really complicated interaction between two different libraries: angular-routing, and single-spa. They both operate with different ideas about how routing is to be accomplished and who is in control of routing.

I'll try to take a look at the repo you created, but if I'm understanding correctly it's only consistently reproducible in IE which will make debugging extremely hard (I use linux as my daily driver). I wouldn't suggest deleting lines of code in single-spa as a solution. I know it's a slow response and I thank you for your continued patience.

**Comment by gubertcalixto** [comment]
@TheMcMurder Actually is totally reproducible in Chrome (using Windows at least). The bug happens 100% in my repo using the following settings:
Chrome v78
Windows 10

**Comment by kylethebaker** [comment]
Also can confirm this happens in several of our angular micro-apps, running both the latest chrome and firefox nightly. It doesn't happen on every navigation event, but when it does happen it will happen consistently for a particular one.

This is something that I'm investigating as well, I'll post here if I figure anything out.

**Comment by kisslove** [comment]
the redirection infinite loop stills exist.any solutions?

**Comment by kisslove** [comment]
I try all solutions above,but failed. so i spend one day to solve the problem.luckily,it is fixed for me.
update file:src/navigation/navigation-events.js
details as my fork: [https://github.com/kisslove/single-spa/blob/master/src/navigation/navigation-events.js](url)

**Comment by gubertcalixto** [comment]
@kisslove Could you explain your [commit](https://github.com/kisslove/single-spa/commit/4e100aefe52fdfabd77bf7e5f38ee525a0589f1a), why adding "isNavGoOrBack" made it work? I think your commit is about go back in browser bug, but actually the route loop can occur in others situations (in my [repo ](https://github.com/gubertcalixto/single-spa-crazy-route)I made an example only using 1 redirect after another)

**Comment by kisslove** [comment]
@gubertcalixto I add ‘isNavGoOrBack‘ ,that just make a distinction between click back-forward in browser and normal route. because the [popstate ](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event) event will be triggered by doing a browser action such as a click on the back or forward button (or calling history.back() or history.forward() in JavaScript).
Normal route will trigger history.pushState and history.replaceState,but go back in browser only trigger  history.replaceState. the problem is history.pushState and history.replaceState  both call urlReroute(createPopStateEvent(state)).obsolutely,it duplicates.

**Comment by kisslove** [comment]
@gubertcalixto I test your [repo](https://github.com/gubertcalixto/single-spa-crazy-route/blob/master/test-app/src/app/two/two.component.ts#L16),it works after using my update.

**Comment by pankajparkar** [comment]
This issue only happens if user click on two different single spa link in a fraction of second. To fix this issue I disabled the links until any of the single page application is loading.

When any application loads via single-spa, it runs below events at start and end.
start - single-spa:before-routing-event
end - single-spa:routing-event
In between there are other events also gets called.

```
@HostListener('window:single-spa:before-routing-event')
beforeRouting() {
    const navbar: HTMLHtmlElement = document.querySelector('navbar');
    const activeLink: HTMLHtmlElement = document.querySelector('.nav-link.active')
    navbar.style.pointerEvents = 'none';
    activeLink.style.classList.add('loading');
}

@HostListener('window:single-spa:routing-event')
afterRoutingEvent() {
    setTimeout(_ => {
      const navbar: HTMLHtmlElement = document.querySelector('navbar');
      const activeLink: HTMLHtmlElement = document.querySelector('.nav-link.active')
      navbar.style.pointerEvents = 'none';
      activeLink.style.classList.remove('loading');
    }, 500);
}
```

**Comment by gubertcalixto** [comment]
@kisslove maybe you could create an PR with your correction for single spa. The bugs that I know/heard about was:
1. Reproduced in my [repo](https://github.com/gubertcalixto/single-spa-crazy-route) (fast change of route or using route promises)
2. The bug you tested (navigate back/forward using browser actions)
3. Another one that happened because of multiple angular redirects in production because of an router guard (but not sure if this one is the same of the tested in my repo).

**PS:** If you don't do it I probably will do in the next month (on vacation right now 😎)

**Comment by kylethebaker** [comment]
@pankajparkar The cases where I see it is when the back button is pressed or when doing a single direct navigation from using the router (not triggered by user interaction). It's possible though that either of those are causing cascading navigation events (guards, redirects, etc), so the theory of "too many navigation's too quickly" is valid.

I'm using the fork by @kisslove and it appears to be working, I did notice one time where the redirects started for a couple of bounces and then resolved themselves quickly, but that was single time out of several. So far I haven't seen any other negative side-effects or regressions within my apps from it.

**Comment by kisslove** [comment]
@gubertcalixto , @kylethebaker .Notice that my pre updates could lead to some display problems when it routes between multi angular micro-fronts.
Now,my solution is that adding a setTimeout fn to solve the loop.it also appears to be working. 
only update [here](https://github.com/CanopyTax/single-spa/blob/master/src/navigation/navigation-events.js#L110).

```
window.history.replaceState = function (state) {
    const result = originalReplaceState.apply(this, arguments);
    // delete the row:
    //urlReroute(createPopStateEvent(state));
    //update as follow:
    setTimeout(() => {
        urlReroute(createPopStateEvent(state));
    }, 50);
    return result;
};
```

**Comment by kylethebaker** [comment]
@kisslove with the setTimeout it's worse for me, I get stuck in the loop but it recovers after a few seconds (so no longer infinite but still won't work for us). I've noticed with the original change that I still do a really quick back-and-forth redirect when it happens, it just stops early enough that it's not very noticeable.

I'm also digging around with the navigation trying to see if I can get to the bottom of it. It does seem like there might be some kind of timing issue at hand though.

**Comment by pankajparkar** [comment]
Our application bundles are quite heavy, they take seconds to load, and when these bundles are downloading and bootstrapping the application, if user try to switches between applications, redirection get into infinite loop. 

The solution I applied is based on my assumption seems to be working for me. It could have been wrong in this approach. Please suggest if there is better way to fix this issue.

**Comment by pankajparkar** [comment]
@kylethebaker I will give a shot with @kisslove 's solution. Just curious to know, after updating the single-spa code, how you're utilising this modified package. From forked git repository, or you have hosted this package somewhere else?

**Comment by kylethebaker** [comment]
@pankajparkar We have a private npm package registry that we use for internal packages, I forked it and published there.

There might be a simpler way to do it, but I updated the package name to `@our-namespace/single-spa` and also updated the `typings.d.ts` to reflect this new name, then updated the consumer app's package.json and imports to use the new package.

**Comment by pankajparkar** [comment]
@kylethebaker Thanks for sharing the way.
I'm afraid doing that, because in future `single-spa` releases, I won't be receiving any updates, Every update has to done manually.

**Comment by joeldenning** [comment]
Perhaps releated - https://github.com/single-spa/single-spa/issues/438#issue-564495428

**Comment by gubertcalixto** [comment]
After a long time debugging, found out that, at least the bug that is in my repo, occurs due to this [line](https://github.com/single-spa/single-spa/blob/master/src/navigation/reroute.js#L113). 
I still could not found the reason of the problem but may have some ideas / conclusions (**I AM NOT 100% WHAT I'M SAYING**):

1. When Angular navigates to an route using a promise which gets in a **catch**, single-spa receives a replaceState using the same route (before the redirect that fails).
- If was in **/one** and failed to **/two**, Angular send a replaceState to **/one**
2. It seems that single-spa calls for the same promise that just happened in Angular (doing a loop between routes **/one** and **/two**).
3. Comment this [line](https://github.com/single-spa/single-spa/blob/master/src/navigation/reroute.js#L113) or doing an if in this [line](https://github.com/single-spa/single-spa/blob/master/src/navigation/navigation-events.js#L102), only activating it` if arguments[2] !== location.pathname`, being arguments[2] equal to the redirect path

**Comment by joeldenning** [comment]
I debugged this a while ago and came to the same conclusion @gubertcalixto. My best guess is that Angular Router relies on all navigation events occurring via the router and within the same Zone. In single-spa, other applications sometimes cause navigation events which causes problems for Angular Router. Also, even within a single application the problem can occur because single-spa delays the `popstate` event from firing to all the listeners on the page until it knows whether to unmount the applications in charge of those listeners or not. This hasn't been a problem for any framework except for Angular, and my best guess is that it is related to the popstate event occurring outside of the Zone or outside of the Angular Router instance.

^ This is all the information I have. It sounds like you've made a ton of progress. If you get to a point where you have a suggested fix, that would be amazing. This issue has plagued single-spa-angular for a long time.

**Comment by joeldenning** [comment]
From https://github.com/single-spa/single-spa/issues/438:

> Could you please provide me some context on why do we override window event listener functions and capturing popstate and hashchange events and processing them. Or can you redirect me to any documentation around single spa architecture. It would be of great help if you provide this information.

Sure! Single-spa overrides window.addEventListener and window.removeEventListener to prevent single-spa applications that are about to be unmounted from receiving a [PopStateEvent](https://developer.mozilla.org/en-US/docs/Web/API/PopStateEvent) or [HashChangeEvent](https://developer.mozilla.org/en-US/docs/Web/API/HashChangeEvent). Since single-spa applications are only aware of the routes described by their activity functions, they would show a Not Found 404 page if we let the PopStateEvent / HashChangeEvent through to them before we unmount them.

In other words, single-spa needs to be the "first" handler of the popstate and hashchange events. And its handling of those events is asynchronous, since it calls `bootstrap()`, `mount()`, and `unmount()` during its handling of those events. Since it needs to be (1) first, and (2) asynchronous, we withhold the routing events from all single-spa applications until it finishes. The patching of `window.addEventListener` is the only way that I'm aware of to accomplish that.

Here are some links in the single-spa source code that might shed further light on how it's happening:

- https://github.com/single-spa/single-spa/blob/master/src/navigation/navigation-events.js#L82-L88
- https://github.com/single-spa/single-spa/blob/master/src/navigation/reroute.js#L39
- https://github.com/single-spa/single-spa/blob/master/src/navigation/reroute.js#L122
- https://github.com/single-spa/single-spa/blob/700eb319184146d3594179224a073ec1bab64133/src/navigation/navigation-events.js#L71
- Note that during "reroute()", unmounted applications call `window.removeEventListener('popstate')`, which is what is needed for them not to show a Not Found 404 page once callCapturedEventListeners is invoked.

It's worth noting that no other frameworks have issue with this approach - react, vue, and others all are fine with it. My guess for the last while is that ZoneJS / Angular Router are assuming a specific timing or zone for the popstate and hashchange events that does not always happen when single-spa withholds the event for a short time before firing the Angular Router popstate event handler.

**Comment by muhlba91** [comment]
@joeldenning @gubertcalixto we are facing the same issue. Is there any progress or are there ideas on how to get it fixed? :)

**Comment by joeldenning** [comment]
Hi @muhlba91 - as I've said above, I would appreciate outside contribution on this issue.

Angular router and ZoneJS are extremely complex (I've read their source code!) and have been difficult for me to debug. I personally do not use Angular very often and single-spa-angular is often a source of pain and frustration for me, since I find Angular's approach (with schematics, hidden webpack config, and Angular CLI) to be quite brittle. I often do not prioritize spending time on single-spa-angular simply because I do not enjoy it.

I recently quit my job to do open source + consulting full time. I may get to this issue at some point, but will likely focus on Angular 9 support. I would gladly prioritize working on this issue if sponsored (Twitter @joelbdenning). Otherwise, I'll look at this issue eventually on a good day when I'm mentally prepared to spend several hours reading minified angular router code 😄

**Comment by muhlba91** [comment]
Yeah, I totally see your point here and I’m not deep into frontend development myself and prefer more lightweight frameworks like vue and debugging such code is not my favorite thing to spend my days with. 😄 (And kudos for reading through both libraries and your open answer is really appreciated! 👍)

Well, business needs and companies go for various reason for Angular which is why I ended up digging through that as well today while trying to debug the problem. 🙈

I’m willing to spend some time/days on trying to at least find a working „hack/workaround“ which could then, hopefully, also open possibilities for a fix. 😃
However, as you have said already, the way all of those libraries work together is quite complex and I was hoping for any direction or idea you have got where I could start to pin down the problem? (Whereas I am not entirely sure about the outcome considering your expertise is way better than mine, tbh now.)

**Comment by joeldenning** [comment]
> where I could start to pin down the problem?

See https://github.com/single-spa/single-spa-angular/issues/113#issuecomment-592981422 and https://github.com/single-spa/single-spa/releases/tag/v5.2.0

**Comment by joeldenning** [comment]
Also note that this is almost certainly related to https://github.com/single-spa/single-spa-angular/issues/94

**Comment by joeldenning** [comment]
I believe that this is fixed in https://github.com/single-spa/single-spa-angular/releases/tag/v3.3.1 and https://github.com/single-spa/single-spa-angular/releases/tag/v4.0.0. single-spa-angular@3 is for Angular 2-8, and single-spa-angular@4 is for Angular 9. See #94 for the explanation and links to PRs that fixed it.

If someone could confirm that this is fixed in 3.3.1 and 4.0.0, please let me know here so I can close this issue.

**Comment by white-collar** [comment]
> If someone could confirm that this is fixed in 3.3.1 and 4.0.0, please let me know here so I can close this issue.

I can confirm - Issue is fixed

**Comment by joeldenning** [comment]
Great! Big thanks to @arturovt for finding and fixing this

**Comment by OriolInvernonLlaneza** [comment]
> > If someone could confirm that this is fixed in 3.3.1 and 4.0.0, please let me know here so I can close this issue.
> 
> I can confirm - Issue is fixed

Hi @white-collar . Did you confirm it in 3.3.1 or 4.0? I'm on the process of updating our Angular 8 micros to 3.3.1 and If I navigate too fast between two of them I still get the infinite loop.

**Comment by matt-gold** [comment]
I am seeing the infinite redirect bug after clicking the back button between two single-spa-angular 4.0 apps, using the latest single-spa 5.3.4.  The bug does not appear if I use @kisslove 's fork of single-spa mentioned in this thread, which is what we've been using as a workaround for this bug for the past few months.

There are a lot of moving parts in my application, but I'll try to get a repo standing that concisely reproduces the issue.

More background on my app which is admittedly not the intended/recommended setup - rather than have single-spa manage the lifecycle of my angular apps, I have a root Angular application that loads other single-spa-angular apps at certain routes as parcels. I noticed that with single-spa-angular 4.0 there've been some changes to better isolate the zones when multiple angular apps are running together. namely the monkey patch of NgZone.isInAngularZone and injecting the zone into the SingleSpaPlatformLocation. I figured I should do the same thing in my root Angular app to try to fix this issue, unfortunately it doesn't seem to make a difference. Here is the bootstrap code from my root app for reference (much of which is lifted from single-spa-angular source):

```
  const appIdentifier = 'myAppRoot';
  NgZone.isInAngularZone = () => {
    // @ts-ignore
    return window.Zone.current._properties[appIdentifier] === true;
  };
  const module = platformBrowserDynamic(getSingleSpaExtraProviders())
    .bootstrapModule(AppModule);

  /// cast to any because SingleSpaPlatformLocation is not exported
  const rootPlatformLocation = module.injector.get(PlatformLocation) as any;
  const rootZone = module.injector.get(NgZone);

  // tslint:disable-next-line:no-string-literal
  rootZone['_inner']._properties[appIdentifier] = true;
  rootPlatformLocation.setNgZone(rootZone);
```

**Comment by varora1406** [comment]
I also have a situation, but with different code.
recreated that situation with https://github.com/varora1406/coexisting-angular-microfrontends/commit/7073a82e237d47fbb809beb3e12a861c2d91c4d3.

Any helpful comment from community on it?

**Comment by joeldenning** [comment]
To solve this bug, we need Angular router to ignore [the single-spa popstateevent](https://single-spa.js.org/docs/api#popstateevent) if it was the app that triggered the navigation by calling `pushState()`.

@arturovt and I have discussed this but haven't found a solution for it. We have both shifted our attention to other work right now, but would gladly accept a PR for anyone who has an idea of how to accomplish the fix described above.

**Comment by joeldenning** [comment]
Another thing - the single-spa core team is proposing a new dom event called `statechange` that would better facilitate multiple SPA routers on the page at the same time. The root of this URL infinite redirect bug is Angular's handling of multiple routers on the same page, and that statechange event would provide a clearer path for Angular's code to handle this properly. The proposal can be found here - https://github.com/whatwg/html/issues/5562

^ Sharing that not because it will end up being the fix we use immediately, but just that we're trying to make progress on all fronts with this bug. Angular Router is a rather brittle router in that it expects no other code to ever trigger navigation events. However, we'll continue look for a workaround that.

**Comment by joeldenning** [comment]
Here's the proposed workaround:

- If a popstate event was caused by this application calling pushState, ignore it.
- Otherwise, re-render by calling this.ngZone.run()

Here's an idea of what the code might be for it:

```
class SingleSpaPlatformLocation extends ɵBrowserPlatformLocation {
  private skipNextPopState: boolean;
  pushState(state: any, title: string, url: string): void {
    this.skipNextPopState = true
    return super.pushState(state, title, url);
  }
  replaceState(state: any, title: string, url: string): void {
    this.skipNextPopState = true
    return super.replaceState(state, title, url);
  }
  onPopState(fn: (event: LocationChangeEvent) => void): void {
    super.onPopState(event => {
      if (this.skipNextPopState) {
        this.skipNextPopState = false;
      } else {
        this.ngZone.run(() => fn(event));
      }
    });
  }
}
```

**Comment by joeldenning** [comment]
The fix for this is released for Angular 9 projects under the `alpha` dist tag. See https://github.com/single-spa/single-spa-angular/releases/tag/v4.1.0-alpha.0.

We plan on backporting the fix to single-spa-angular@3, as shown in #192. I'll post here once a new 3.x is released.

Could people please try out 4.1.0-alpha.0 and let us know if it fixes the issue for them? We'd like to get as many people verifying that it's fixed as possible.

**Comment by joeldenning** [comment]
One note about the fix:

**You must upgrade to single-spa@>=5.4.0** for the fix to work!

**Comment by joeldenning** [comment]
This fix has been backported for Angular 2-8 via https://github.com/single-spa/single-spa-angular/releases/tag/v3.4.0. Note that you must use single-spa@>=5.4.0, and add the extra single-spa providers to your code in order for it to work.

Please comment here after having tried it out to let us know if it's working for everyone.

**Comment by gubertcalixto** [comment]
Tried to use the v3.4.1 in my repo and unfortunately had the following error
![image](https://user-images.githubusercontent.com/28396856/82605208-f1efb780-9b8b-11ea-8a17-7d11c6969d79.png)
Thing I did before test it:

1. Update single-spa.min.js import from
System.import('https://cdnjs.cloudflare.com/ajax/libs/single-spa/4.3.5/system/single-spa.min.js')
to 
System.import('https://cdnjs.cloudflare.com/ajax/libs/single-spa/5.5.0/system/single-spa.min.js')

2. Updated single-spa-angular package to 3.4.1
3. Updated main.single-spa.ts file as indicated in release notes

**Comment by joeldenning** [comment]
This looks like an error with differential loading (es5 vs es2015). single-spa-angular@3 doesn't use ng-packagr right now - @arturovt can you comment on what our options are to support differential loading for single-spa-angular@3?

@gubertcalixto what you can try is to use the es2015 version of your bundled output while testing the bug - that one will call classes with `new` correctly

**Comment by gubertcalixto** [comment]
Test it over again, this time building the app and accessing **main-es5.js** and it **worked just fine** (main-es2015.js did not work, with same error as above). 
Anyway, this button below finally does not make infinite loop anymore.  
![image](https://user-images.githubusercontent.com/28396856/82607896-1057b200-9b90-11ea-877b-1161c5a183fe.png)

THANK YOU GUYS 👍

**Comment by arturovt** [comment]
I was testing so fast that I forgot about differential loading.... :grimacing: 

I will think about what we can do with that.

**Comment by varora1406** [comment]
For me as well, this fix of `alpha` release is good till now.

**Comment by joeldenning** [comment]
For those using single-spa-angular@3, we released https://github.com/single-spa/single-spa-angular/releases/tag/v3.5.0 which has support for differential loading. With all the feedback positive so far on the 4.x alpha release, I think we should just release this as latest. @arturovt do you agree?

**Comment by arturovt** [comment]
I would want to hear feedback from @matt-gold since he was very active in helping to test out our code changes.

**Comment by joeldenning** [comment]
This fix is published under the `latest` dist-tag. Release notes at https://github.com/single-spa/single-spa-angular/releases/tag/v4.2.0.

**Comment by matt-gold** [comment]
@arturovt 
The bug is fixed for me on single-spa-angular 4.2.0 + single-spa 5.4.2 👍

**Comment by varora1406** [comment]
Was working on this issue, it's reproducible.
Interesting fact related to better situation is also attached in code
https://github.com/varora1406/coexisting-angular-microfrontends/tree/issue-inifnite-url-loop

**Comment by joeldenning** [comment]
Reopening this while we investigate the repo provided by @varora1406.

**Comment by arturovt** [comment]
@varora1406 you didn't provide any steps to reproduce it. What do I have to do?

**Comment by varora1406** [comment]
@arturovt sorry for that.
just try to switch randomly with routes fast. You will get the results in a minute or two most probably.

**Comment by varora1406** [comment]
I have also added a button in latest commit, which can help you recreate bug easily

**Comment by joeldenning** [comment]
Someone contacted me in the single-spa slack workspace this morning saying that they had found a change that might work as a solution. The change is to modify `skipNextPopState` to be a number that acts as a counter - every time pushState/replaceState is called, it is incremented; and every time that popstate is fired it is decremented. He said that it's possible for multiple pushStates and replaceStates to pile up while waiting for the popstate event.

**Comment by varora1406** [comment]
@joeldenning no pull request is opened yet for this issue.
can you share implementation approach details?

**Comment by joeldenning** [comment]
Hi @varora1406, my previous comment details the implementation. I'll add a few more links for anyone who's interested in submitting a pull request with the change:

https://github.com/single-spa/single-spa-angular/blob/a4cc1633bf0b8f39c57484a51e7895561818c37b/src/src/extra-providers.ts#L15

https://github.com/single-spa/single-spa-angular/blob/a4cc1633bf0b8f39c57484a51e7895561818c37b/src/src/extra-providers.ts#L54-L55

https://github.com/single-spa/single-spa-angular/blob/a4cc1633bf0b8f39c57484a51e7895561818c37b/src/src/extra-providers.ts#L36-L37

https://github.com/single-spa/single-spa-angular/blob/a4cc1633bf0b8f39c57484a51e7895561818c37b/src/src/extra-providers.ts#L63

The `skipNextPopState` variable should be changed to an integer instead of a boolean. It should be incremented when pushState/replaceState are called, and decremented whenever `onPopState` is called. We should only call `ngZone.run()` when it is `0`

I'd gladly review a PR that implements the change.

**Comment by arturovt** [comment]
It'd also be great if we reproduce that behavior in our tests before making that change. And after the change is done we'll be able to see if that code change really fixes that issue.

**Comment by joeldenning** [comment]
Here's the conversation from Slack about this.

![image](https://user-images.githubusercontent.com/5524384/86700214-2a2d4700-bfce-11ea-84ad-96ffd0df4cef.png)

I don't know Dmitry's Github handle, but if you're watching this thread Dmitry could you please comment on whether your solution has worked?

**Comment by DmitryBrus** [comment]
@joeldenning 
I didn't observe infinite loop, only one redundant redirection.
Before opening PR I'd like to be sure that I can clearly reproduce the issue, unfortunately I do not see it happen on new  repo, I tried and played around, set two redirection close one another but still no luck.
The only place it happen for me in our pretty big app, I tried to remove one by one all irrelevant parts and see if I can get clear reproduction... still in progress.

**Comment by varora1406** [comment]
@DmitryBrus you can try it at https://github.com/varora1406/coexisting-angular-microfrontends/tree/issue-inifnite-url-loop
I am also going to start work from today after office hours on this to fix 👩‍🚀

**Comment by gauravpandvia** [comment]
@joeldenning @arturovt After the version 3.4.1 fix of infinite redirection, it came back haunting again in our app. But this time it was weird, as it was not happening consistently but intermittently. Seeing the above comments as @DmitryBrus said it is happening when there are two close redirection routing in the angular app. Since in a more complex big app, it is quite visible and pertinent but I tried creating a small reproducible version of the problem here in this cloned template of [coexisting-angular-frontend repo](https://github.com/gauravpandvia/singlespa-angular). 

In **/app1** I have used Angular v8 and single-spa-angular v3.6.0 and added a route named **/test1** & **/test2** in it. I have provided the link to Test1 Component in app.component.html and in Test1 component I have added a router navigation to Test2 component under ngOnInit. Observe that it happens only in the first time of app start, that when you will click on Test 1 it will go to test2 then back to test 1 then back to test2. Also, if you put a debug point on ngOnInit router navigate command in Test1 Component, you will notice that it is being paused  infinite times. Hope this helps and we can get a solution of this.

**Comment by joeldenning** [comment]
Thanks for the repo demonstrating the issue. I do not plan on working on this issue, but welcome anybody who does to discuss it here and/or submit a pull request.

**Comment by plrthink** [comment]
Any updates on this? I'm still facing this issue which causes infinite loop between two routes.

**Comment by ashuaggarwal94** [comment]
@plrthink  You can try  imports: [RouterModule.forRoot(routes, { onSameUrlNavigation: "ignore" })], in angular routing module it worked for me!! maybe you can give it a shot!!!

**Comment by plrthink** [comment]
Well, I don't think it would work since 'ignore' is the default value of the option onSameUrlNavigation.

**Comment by plrthink** [comment]
And it's indeed not working after testing.

**Comment by arturovt** [comment]
Hey guys, we have released the fix in 4.6.0 version. Could anyone check if it resolves your issue? @varora1406 could you also check in your example?

**Comment by joeldenning** [comment]
Closing since the fix is published in 4.6.0 - we can reopen if that fix doesn't solve all cases.

**Comment by JudahMorrison** [comment]
Hello, single spa crew.
I have been dealing with this issue for a while and upgrading single-spa-angular to 4.7 and single-spa to 5.8.1 didn't seem to fix it for me.

**Comment by arturovt** [comment]
> Hello, single spa crew.
> I have been dealing with this issue for a while and upgrading single-spa-angular to 4.7 and single-spa to 5.8.1 didn't seem to fix it for me.

Could you paste your main.single-spa.ts?

**Comment by JudahMorrison** [comment]
import { enableProdMode, NgZone } from '@angular/core';

import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { Router } from '@angular/router';
import { ɵAnimationEngine as AnimationEngine } from '@angular/animations/browser';

import { singleSpaAngular, getSingleSpaExtraProviders } from 'single-spa-angular';


import { AppModule } from './app/app.module';
import { environment } from './environments/environment';
import { singleSpaPropsSubject } from './single-spa/single-spa-props';

if (environment.production) {
  enableProdMode();
}

const lifecycles = singleSpaAngular({
  bootstrapFunction: singleSpaProps => {
    singleSpaPropsSubject.next(singleSpaProps);
    console.log('BOOTSTRAP FUNCTION LANDING');
    return platformBrowserDynamic(getSingleSpaExtraProviders()).bootstrapModule(AppModule);
  },
  template: '<app-root-landing />',
  Router,
  NgZone,
  AnimationEngine,
});

export const bootstrap = lifecycles.bootstrap;
export const mount = lifecycles.mount;
export const unmount = lifecycles.unmount;

**Comment by arturovt** [comment]
I see, could you add NavigationStart to options in the same way you pass NgZone and Router?

**Comment by JudahMorrison** [comment]
@arturovt You are truly an everyday superhero.

**Comment by kmiasko** [comment]
@arturovt Doesn't resolve this issue for our team too. For easier reproduction try setting CPU Throttling in Performance tab in developer settings.

I've also cloned @varora1406 repo, updated packages, added NavigationStart to start method and i can still reproduce this.

**Comment by arturovt** [comment]
@kmiasko 

I've just cloned @varora1406 repo, updated packages, added NavigationStart to options and I can't reproduce it anymore:

![image](https://user-images.githubusercontent.com/7337691/99805304-fbf77080-2b44-11eb-8234-780035898ffe.png)

Could you provide a minimal reproducible example?

**Comment by kmiasko** [comment]
@arturovt Are you sure you're using issue-inifnite-url-loop branch?

**Comment by varora1406** [comment]
Hi guys, sorry I have been away for some time.

@arturovt I can't see button added by me in your screenshot. Please confirm if you are checking at `issue-infinite-url-loop` branch.
![image](https://user-images.githubusercontent.com/24986539/99870676-12f89a00-2bfb-11eb-8911-e1079133656d.png)

**Comment by kmiasko** [comment]
@arturovt I've added some logging to single-spa-angular app and single-spa-angular lib, maybe you'll see any pattern there.

[routing log](https://gist.github.com/kmiasko/60c76f2c890d3fc800130056708e220c)

**Comment by kmiasko** [comment]
@arturovt My bad. I did everything again from the beginning, by the book on `issue-inifnite-url-loop` branch on @varora1406 example and it's working, there is no loop.

I've also solved our team issue with the loop (and i think back button issue - to be testes carefully). As we used angular sub app inside angular main app i had to also patch router in the main app using `getSingleSpaExtraProviders()` in `platformBrowserDynamic`.

Thank you for your time, and again sorry for the fuzz.

**Comment by arturovt** [comment]
@kmiasko I'm glad that you've solved it!

**Comment by chenji336** [comment]
single-spa-anggular 4.x this question.how to fix infinite loop by single-spa-angular 3.x ?

**Comment by joeldenning** [comment]
single-spa-angular@3 doesn't have it fully fixed - see https://github.com/single-spa/single-spa-angular/pull/235/files#diff-d41e9b77a99ac86d9908fc97b9df8578479c310a0fe85f632f375b2d9e95f39cR202. We need to make a corresponding change in the 3.x branch to fix it. Would you be interested in submitting a pull request with it?

**Comment by nicknasirpour** [comment]
Hi @chenji336 @joeldenning I will submit a pull request sometime this week implementing the fix for the 3.x branch.

Something of note is that this solution in single-spa-angular alone won't solve 100% of issues.

I have a base project, which is an angular module, where singleSpa.start() is called. My parcels are subsequently mounted onto a component of this project. In the base project, during the redirect from my login component to another component, I was getting this infinite loop. The infinite loop behaviour stopped when I removed the call to singleSpa.start(). I had to implement the non-imperative skipping logic in my base application as well.

**Comment by thanhbinhtran93** [comment]
Adding `Router`, `NavigationStart` and `getSingleSpaExtraProviders` like this would help for my case when using redirectTo in RouterModule:

```
const lifecycles = singleSpaAngular({
  bootstrapFunction: (singleSpaProps) => {
    singleSpaProps$.next(singleSpaProps);
    return platformBrowserDynamic().bootstrapModule(AppModule);
    return platformBrowserDynamic(getSingleSpaExtraProviders()).bootstrapModule(AppModule);
  },
  template: '<app-root />',
  Router,
  NavigationStart,
  NgZone,
});
```
It is also mentioned in docs https://single-spa.js.org/docs/ecosystem-angular/#api-updates
